File: ios/CaddieAI/Services/GolfLogicEngine.swift
//
//  GolfLogicEngine.swift
//  CaddieAI
//

import Foundation

// MARK: - Supporting Types

struct TargetStrategy: Codable, Sendable {
    var target: String
    var preferredMiss: String
    var reasoning: String
}

struct DeterministicAnalysis: Codable, Sendable {
    var effectiveDistanceYards: Int
    var recommendedClub: Club
    var alternateClub: Club?
    var targetStrategy: TargetStrategy
    var adjustments: [String]
    var maxClubForLie: Club?
    var executionPlan: ExecutionPlan
}

// MARK: - Golf Logic Engine

struct GolfLogicEngine {

    // MARK: - Primary Entry Point

    static func analyze(
        context: ShotContext,
        profile: PlayerProfile
    ) -> DeterministicAnalysis {
        let effectiveDistance = calculateEffectiveDistance(context: context, ironType: profile.ironType)
        let clubLimit = maxClubForLie(context.lieType, ironType: profile.ironType)
        let recommendedClub = selectClub(
            effectiveDistance: effectiveDistance,
            clubDistances: profile.clubDistances,
            maxAllowedClub: clubLimit
        )
        let alternateClub = selectAlternateClub(
            effectiveDistance: effectiveDistance,
            primaryClub: recommendedClub,
            clubDistances: profile.clubDistances,
            maxAllowedClub: clubLimit
        )
        let targetStrategy = determineTargetStrategy(
            context: context,
            profile: profile,
            recommendedClub: recommendedClub
        )
        let adjustments = describeAdjustments(context: context, ironType: profile.ironType)

        let executionPlan = ExecutionEngine.generateExecutionPlan(
            context: context,
            club: recommendedClub,
            effectiveDistance: effectiveDistance,
            profile: profile
        )

        return DeterministicAnalysis(
            effectiveDistanceYards: effectiveDistance,
            recommendedClub: recommendedClub,
            alternateClub: alternateClub,
            targetStrategy: targetStrategy,
            adjustments: adjustments,
            maxClubForLie: clubLimit,
            executionPlan: executionPlan
        )
    }

    // MARK: - Effective Distance Calculation

    static func calculateEffectiveDistance(context: ShotContext, ironType: IronType? = nil) -> Int {
        var distance = Double(context.distanceYards)

        distance += windAdjustment(
            strength: context.windStrength,
            direction: context.windDirection,
            baseDistance: context.distanceYards
        )

        distance += lieAdjustment(lieType: context.lieType)

        distance += slopeAdjustment(
            slope: context.slope,
            baseDistance: context.distanceYards
        )

        // GI/SGI irons lose distance from challenging lies due to wide soles
        if let ironType {
            distance += ironTypeLieAdjustment(lieType: context.lieType, ironType: ironType)
        }

        return max(1, Int(distance.rounded()))
    }

    // MARK: - Wind Adjustment

    static func windAdjustment(
        strength: WindStrength,
        direction: WindDirection,
        baseDistance: Int
    ) -> Double {
        let factor: Double
        switch strength {
        case .none: return 0
        case .light: factor = 0.03
        case .moderate: factor = 0.07
        case .strong: factor = 0.12
        }

        let adjustment = Double(baseDistance) * factor

        switch direction {
        case .into: return adjustment
        case .helping: return -adjustment
        case .crossLeftToRight, .crossRightToLeft:
            return adjustment * 0.3
        }
    }

    // MARK: - Lie Adjustment

    static func lieAdjustment(lieType: LieType) -> Double {
        switch lieType {
        case .fairway: return 0
        case .firstCut: return 3
        case .rough: return 7
        case .deepRough: return 15
        case .greensideBunker: return 5
        case .fairwayBunker: return 10
        case .hardpan: return -3
        case .pineStraw: return 5
        case .treesObstructed: return 20
        }
    }

    // MARK: - Slope Adjustment

    static func slopeAdjustment(slope: Slope, baseDistance: Int) -> Double {
        let base = Double(baseDistance)
        switch slope {
        case .level: return 0
        case .uphill: return base * 0.05
        case .downhill: return -(base * 0.05)
        case .ballAboveFeet: return 3
        case .ballBelowFeet: return 3
        }
    }

    // MARK: - Iron Type Lie Adjustment

    /// Additional distance penalty for GI/SGI irons from challenging lies.
    /// Wide soles and high offset make these irons less effective from
    /// bunkers, tight lies, and thick rough.
    static func ironTypeLieAdjustment(lieType: LieType, ironType: IronType) -> Double {
        let multiplier: Double = ironType == .superGameImprovement ? 1.5 : 1.0
        switch lieType {
        case .fairwayBunker:    return 8 * multiplier   // wide sole digs, poor contact
        case .greensideBunker:  return 5 * multiplier   // harder to open face with offset
        case .hardpan:          return 5 * multiplier   // wide sole bounces off hardpan
        case .rough:            return 3 * multiplier   // grass grabs wide sole
        case .deepRough:        return 5 * multiplier   // even worse in heavy rough
        default:                return 0
        }
    }

    // MARK: - Club Selection

    static func selectClub(
        effectiveDistance: Int,
        clubDistances: [ClubDistance],
        maxAllowedClub: Club?
    ) -> Club {
        let sorted = clubDistances
            .filter { isClubAllowed($0.club, maxAllowed: maxAllowedClub) }
            .sorted { $0.carryYards > $1.carryYards }

        // Find the shortest club that still covers the distance
        var bestClub = sorted.last?.club ?? .pitchingWedge

        for cd in sorted {
            if cd.carryYards >= effectiveDistance {
                bestClub = cd.club
            } else {
                break
            }
        }

        return bestClub
    }

    // MARK: - Alternate Club

    static func selectAlternateClub(
        effectiveDistance: Int,
        primaryClub: Club,
        clubDistances: [ClubDistance],
        maxAllowedClub: Club?
    ) -> Club? {
        let sorted = clubDistances
            .filter { isClubAllowed($0.club, maxAllowed: maxAllowedClub) }
            .sorted { $0.club.sortOrder < $1.club.sortOrder }

        guard let primaryIndex = sorted.firstIndex(where: { $0.club == primaryClub }) else {
            return nil
        }

        // Check one club longer and one club shorter
        let longerIndex = primaryIndex - 1
        let shorterIndex = primaryIndex + 1

        if shorterIndex < sorted.count {
            let shorter = sorted[shorterIndex]
            let diff = abs(shorter.carryYards - effectiveDistance)
            if diff <= 10 {
                return shorter.club
            }
        }

        if longerIndex >= 0 {
            let longer = sorted[longerIndex]
            let diff = abs(longer.carryYards - effectiveDistance)
            if diff <= 10 {
                return longer.club
            }
        }

        return nil
    }

    // MARK: - Club Restrictions by Lie

    static func maxClubForLie(_ lieType: LieType, ironType: IronType? = nil) -> Club? {
        // Base limits
        var limit: Club?
        switch lieType {
        case .deepRough: limit = .iron7
        case .greensideBunker: limit = .sandWedge
        case .fairwayBunker: limit = .iron6
        case .treesObstructed: limit = .iron7
        case .pineStraw: limit = .iron5
        default: limit = nil
        }

        // GI/SGI irons need tighter limits from challenging lies
        if let ironType {
            switch lieType {
            case .fairwayBunker:
                // Wide sole makes long irons from bunkers very difficult
                limit = ironType == .superGameImprovement ? .iron8 : .iron7
            case .hardpan:
                // Wide sole bounces — restrict to shorter irons
                limit = ironType == .superGameImprovement ? .iron8 : .iron7
            default:
                break
            }
        }

        return limit
    }

    static func isClubAllowed(_ club: Club, maxAllowed: Club?) -> Bool {
        guard let max = maxAllowed else { return true }
        return club.sortOrder >= max.sortOrder
    }

    // MARK: - Target Strategy

    static func determineTargetStrategy(
        context: ShotContext,
        profile: PlayerProfile,
        recommendedClub: Club
    ) -> TargetStrategy {
        var target = "Center of green"
        var miss = "Safe side of the green"
        var reasoning = ""

        let hasHazardNotes = !context.hazardNotes.trimmingCharacters(in: .whitespaces).isEmpty
        let hazardLower = context.hazardNotes.lowercased()

        // Adjust target based on player tendency and aggressiveness
        switch context.aggressiveness {
        case .conservative:
            target = "Center of the green, away from trouble"
            reasoning = "Conservative approach favors the widest part of the green."
        case .normal:
            target = "Center of the green"
            reasoning = "Standard approach targeting the middle of the green."
        case .aggressive:
            target = "Pin location"
            reasoning = "Aggressive play targeting the flag."
        }

        // Adjust miss side based on miss tendency
        switch profile.missTendency {
        case .right:
            miss = "Favor the left side — your miss tends right"
            if !reasoning.isEmpty { reasoning += " " }
            reasoning += "Accounting for right miss tendency."
        case .left:
            miss = "Favor the right side — your miss tends left"
            if !reasoning.isEmpty { reasoning += " " }
            reasoning += "Accounting for left miss tendency."
        case .thin:
            miss = "Club up to account for thin contact tendency"
            if !reasoning.isEmpty { reasoning += " " }
            reasoning += "Thin misses lose distance."
        case .fat:
            miss = "Club up to account for heavy contact tendency"
            if !reasoning.isEmpty { reasoning += " " }
            reasoning += "Heavy contact loses distance."
        case .straight:
            miss = "No major miss adjustment needed"
        }

        // Adjust for slope-induced shape changes
        let clubShape = profile.stockShapeForClub(recommendedClub)
        switch context.slope {
        case .ballBelowFeet:
            if !reasoning.isEmpty { reasoning += " " }
            reasoning += "Ball below feet promotes a fade."
            if clubShape == .fade {
                miss = "Favor the left side — stance will exaggerate your fade"
            }
        case .ballAboveFeet:
            if !reasoning.isEmpty { reasoning += " " }
            reasoning += "Ball above feet promotes a draw."
            if clubShape == .draw {
                miss = "Favor the right side — stance will exaggerate your draw"
            }
        default:
            break
        }

        // Adjust for wind crosswind
        switch context.windDirection {
        case .crossLeftToRight:
            if !reasoning.isEmpty { reasoning += " " }
            reasoning += "Crosswind L→R will push the ball right."
            target += ", aiming slightly left to allow for wind"
        case .crossRightToLeft:
            if !reasoning.isEmpty { reasoning += " " }
            reasoning += "Crosswind R→L will push the ball left."
            target += ", aiming slightly right to allow for wind"
        default:
            break
        }

        // Factor in hazard notes
        if hasHazardNotes {
            if hazardLower.contains("water left") || hazardLower.contains("hazard left") {
                miss = "Miss right — water left"
                if !reasoning.isEmpty { reasoning += " " }
                reasoning += "Water left makes right the safe miss."
            } else if hazardLower.contains("water right") || hazardLower.contains("hazard right") {
                miss = "Miss left — water right"
                if !reasoning.isEmpty { reasoning += " " }
                reasoning += "Water right makes left the safe miss."
            }
            if hazardLower.contains("bunker short") {
                if !reasoning.isEmpty { reasoning += " " }
                reasoning += "Bunker short of green — make sure to take enough club."
            }
            if hazardLower.contains("ob") || hazardLower.contains("out of bounds") {
                if !reasoning.isEmpty { reasoning += " " }
                reasoning += "OB in play — favor the safe side."
            }
        }

        return TargetStrategy(
            target: target,
            preferredMiss: miss,
            reasoning: reasoning
        )
    }

    // MARK: - Describe Adjustments

    static func describeAdjustments(context: ShotContext, ironType: IronType? = nil) -> [String] {
        var adjustments: [String] = []

        let windAdj = windAdjustment(
            strength: context.windStrength,
            direction: context.windDirection,
            baseDistance: context.distanceYards
        )
        if abs(windAdj) > 0.5 {
            let sign = windAdj > 0 ? "+" : ""
            adjustments.append("Wind (\(context.windStrength.displayName) \(context.windDirection.displayName)): \(sign)\(Int(windAdj.rounded())) yards")
        }

        let lieAdj = lieAdjustment(lieType: context.lieType)
        if abs(lieAdj) > 0.5 {
            let sign = lieAdj > 0 ? "+" : ""
            adjustments.append("Lie (\(context.lieType.displayName)): \(sign)\(Int(lieAdj.rounded())) yards")
        }

        let slopeAdj = slopeAdjustment(slope: context.slope, baseDistance: context.distanceYards)
        if abs(slopeAdj) > 0.5 {
            let sign = slopeAdj > 0 ? "+" : ""
            adjustments.append("Slope (\(context.slope.displayName)): \(sign)\(Int(slopeAdj.rounded())) yards")
        }

        if let ironType {
            let giAdj = ironTypeLieAdjustment(lieType: context.lieType, ironType: ironType)
            if abs(giAdj) > 0.5 {
                adjustments.append("\(ironType.shortName) iron penalty (\(context.lieType.displayName)): +\(Int(giAdj.rounded())) yards")
            }
        }

        if adjustments.isEmpty {
            adjustments.append("No adjustments — clean conditions")
        }

        return adjustments
    }
}

File: mobile-flutter/lib/core/golf/golf_logic_engine.dart
// GolfLogicEngine — KAN-293 (S7.2) Flutter port of
// `ios/CaddieAI/Services/GolfLogicEngine.swift`. Per **ADR 0008**,
// this is the most divergence-heavy of the engines: iOS uses
// proportional wind factors and additive lie penalties, while
// Android uses fixed-yardage tables and multiplicative penalties.
// **The Flutter port follows iOS exclusively** — Android will be
// updated post-cutover to match.
//
// **Pure function from `(context, profile)` to
// `DeterministicAnalysis`.** The engine has no I/O, no logging
// (the iOS native logs a warning on `shotType` / `lieType`
// mismatches; we drop that side effect to keep the engine pure
// — the caller can log if it wants to). Calls into ExecutionEngine
// to populate the `executionPlan` field.
//
// **Iron type penalty:** GI / SGI irons get an additional yardage
// penalty from challenging lies (bunkers, hardpan, rough). The
// math is `multiplier × baseYards` where `multiplier = 1.5` for
// SGI and `1.0` for GI. See `ironTypeLieAdjustment` for the
// per-lie table.

import 'execution_engine.dart';
import 'golf_enums.dart';
import 'shot_context.dart';
import 'target_strategy.dart';

abstract final class GolfLogicEngine {
  GolfLogicEngine._();

  // ── Primary entry point (iOS lines 32-80) ───────────────────────

  static DeterministicAnalysis analyze({
    required ShotContext context,
    required ShotPreferences profile,
  }) {
    final effectiveDistance = calculateEffectiveDistance(
      context: context,
      ironType: profile.ironType,
    );
    final clubLimit =
        maxClubForLie(context.lieType, ironType: profile.ironType);
    final recommendedClub = selectClub(
      effectiveDistance: effectiveDistance,
      clubDistances: profile.clubDistances,
      maxAllowedClub: clubLimit,
    );
    final alternateClub = selectAlternateClub(
      effectiveDistance: effectiveDistance,
      primaryClub: recommendedClub,
      clubDistances: profile.clubDistances,
      maxAllowedClub: clubLimit,
    );
    final strategy = determineTargetStrategy(
      context: context,
      profile: profile,
      recommendedClub: recommendedClub,
    );
    final adjustments = describeAdjustments(
      context: context,
      ironType: profile.ironType,
    );
    final plan = ExecutionEngine.generateExecutionPlan(
      context: context,
      club: recommendedClub,
      effectiveDistance: effectiveDistance,
      profile: profile,
    );

    return DeterministicAnalysis(
      effectiveDistanceYards: effectiveDistance,
      recommendedClub: recommendedClub,
      alternateClub: alternateClub,
      targetStrategy: strategy,
      adjustments: adjustments,
      maxClubForLie: clubLimit,
      executionPlan: plan,
    );
  }

  // ── Effective distance (iOS lines 84-106) ───────────────────────

  static int calculateEffectiveDistance({
    required ShotContext context,
    IronType? ironType,
  }) {
    var distance = context.distanceYards.toDouble();

    distance += windAdjustment(
      strength: context.windStrength,
      direction: context.windDirection,
      baseDistance: context.distanceYards,
    );
    distance += lieAdjustment(context.lieType);
    distance += slopeAdjustment(
      slope: context.slope,
      baseDistance: context.distanceYards,
    );

    if (ironType != null) {
      distance += ironTypeLieAdjustment(
        lieType: context.lieType,
        ironType: ironType,
      );
    }

    final rounded = distance.round();
    return rounded < 1 ? 1 : rounded;
  }

  // ── Wind adjustment (iOS lines 110-131) ─────────────────────────

  /// Returns yards to ADD to the base distance. Positive = play
  /// more club. Negative (helping wind) = play less.
  static double windAdjustment({
    required WindStrength strength,
    required WindDirection direction,
    required int baseDistance,
  }) {
    final double factor;
    switch (strength) {
      case WindStrength.none:
        return 0;
      case WindStrength.light:
        factor = 0.03;
      case WindStrength.moderate:
        factor = 0.07;
      case WindStrength.strong:
        factor = 0.12;
    }

    final adjustment = baseDistance.toDouble() * factor;

    switch (direction) {
      case WindDirection.into:
        return adjustment;
      case WindDirection.helping:
        return -adjustment;
      case WindDirection.crossLeftToRight:
      case WindDirection.crossRightToLeft:
        return adjustment * 0.3;
    }
  }

  // ── Lie adjustment (iOS lines 135-147) ──────────────────────────

  /// Yards to add for the lie. Additive — see ADR 0008.
  static double lieAdjustment(LieType lie) {
    switch (lie) {
      case LieType.fairway:
        return 0;
      case LieType.firstCut:
        return 3;
      case LieType.rough:
        return 7;
      case LieType.deepRough:
        return 15;
      case LieType.greensideBunker:
        return 5;
      case LieType.fairwayBunker:
        return 10;
      case LieType.hardpan:
        return -3;
      case LieType.pineStraw:
        return 5;
      case LieType.treesObstructed:
        return 20;
    }
  }

  // ── Slope adjustment (iOS lines 151-160) ────────────────────────

  static double slopeAdjustment({
    required Slope slope,
    required int baseDistance,
  }) {
    final base = baseDistance.toDouble();
    switch (slope) {
      case Slope.level:
        return 0;
      case Slope.uphill:
        return base * 0.05;
      case Slope.downhill:
        return -(base * 0.05);
      case Slope.ballAboveFeet:
      case Slope.ballBelowFeet:
        return 3;
    }
  }

  // ── GI/SGI iron lie penalty (iOS lines 167-177) ─────────────────

  static double ironTypeLieAdjustment({
    required LieType lieType,
    required IronType ironType,
  }) {
    final multiplier =
        ironType == IronType.superGameImprovement ? 1.5 : 1.0;
    switch (lieType) {
      case LieType.fairwayBunker:
        return 8 * multiplier;
      case LieType.greensideBunker:
        return 5 * multiplier;
      case LieType.hardpan:
        return 5 * multiplier;
      case LieType.rough:
        return 3 * multiplier;
      case LieType.deepRough:
        return 5 * multiplier;
      // ignore: no_default_cases
      default:
        return 0;
    }
  }

  // ── Club selection (iOS lines 181-202) ──────────────────────────
  //
  // **Normal case** (some club covers the distance): the loop walks
  // longest → shortest, overwriting `bestClub` each time it finds
  // a club that still covers, then breaks when it hits one that
  // doesn't. The final value is the SHORTEST club whose carry ≥
  // effective distance — the right answer for golf (you don't
  // crush a 50y shot with a driver).
  //
  // **Edge case** (no club covers, e.g. 300y target with a 245y
  // driver): the loop's first iteration immediately fails the
  // `>=` check and breaks. `bestClub` stays at its default, which
  // mirrors iOS's `sorted.last?.club ?? .pitchingWedge` — the
  // shortest allowed club, or a pitching wedge for an empty bag.
  // The empty-bag wedge fallback is clearly intentional in iOS;
  // the "distance exceeds bag" path is an incidental side effect
  // of that choice rather than an explicit recommendation. Per
  // ADR 0008 we replicate iOS exactly. If product later wants a
  // "longest available club" behavior for unreachable targets,
  // that's a deliberate Flutter-side improvement that needs an
  // ADR amendment AND a matching update on the iOS side to keep
  // the byte-identical contract intact.
  static Club selectClub({
    required int effectiveDistance,
    required Map<Club, int> clubDistances,
    Club? maxAllowedClub,
  }) {
    final entries = clubDistances.entries
        .where((e) => isClubAllowed(e.key, maxAllowedClub))
        .toList()
      ..sort((a, b) => b.value.compareTo(a.value));

    if (entries.isEmpty) return Club.pitchingWedge;

    // Default = shortest allowed club. Mirrors iOS's
    // `sorted.last?.club ?? .pitchingWedge`. See file comment above.
    var bestClub = entries.last.key;
    for (final entry in entries) {
      if (entry.value >= effectiveDistance) {
        bestClub = entry.key;
      } else {
        break;
      }
    }
    return bestClub;
  }

  // ── Alternate club (iOS lines 206-241) ──────────────────────────

  static Club? selectAlternateClub({
    required int effectiveDistance,
    required Club primaryClub,
    required Map<Club, int> clubDistances,
    Club? maxAllowedClub,
  }) {
    final allowed = clubDistances.entries
        .where((e) => isClubAllowed(e.key, maxAllowedClub))
        .toList()
      ..sort((a, b) => a.key.sortOrder.compareTo(b.key.sortOrder));

    final primaryIndex =
        allowed.indexWhere((e) => e.key == primaryClub);
    if (primaryIndex < 0) return null;

    final shorterIndex = primaryIndex + 1;
    final longerIndex = primaryIndex - 1;

    if (shorterIndex < allowed.length) {
      final shorter = allowed[shorterIndex];
      if ((shorter.value - effectiveDistance).abs() <= 10) {
        return shorter.key;
      }
    }

    if (longerIndex >= 0) {
      final longer = allowed[longerIndex];
      if ((longer.value - effectiveDistance).abs() <= 10) {
        return longer.key;
      }
    }

    return null;
  }

  // ── Lie-based max-club restrictions (iOS lines 245-272) ─────────

  static Club? maxClubForLie(LieType lie, {IronType? ironType}) {
    Club? limit;
    switch (lie) {
      case LieType.deepRough:
        limit = Club.iron7;
      case LieType.greensideBunker:
        limit = Club.sandWedge;
      case LieType.fairwayBunker:
        limit = Club.iron6;
      case LieType.treesObstructed:
        limit = Club.iron7;
      case LieType.pineStraw:
        limit = Club.iron5;
      // ignore: no_default_cases
      default:
        limit = null;
    }

    if (ironType != null) {
      switch (lie) {
        case LieType.fairwayBunker:
          limit = ironType == IronType.superGameImprovement
              ? Club.iron8
              : Club.iron7;
        case LieType.hardpan:
          limit = ironType == IronType.superGameImprovement
              ? Club.iron8
              : Club.iron7;
        // ignore: no_default_cases
        default:
          break;
      }
    }
    return limit;
  }

  static bool isClubAllowed(Club club, Club? maxAllowed) {
    if (maxAllowed == null) return true;
    return club.sortOrder >= maxAllowed.sortOrder;
  }

  // ── Target strategy (iOS lines 281-387) ─────────────────────────

  static TargetStrategy determineTargetStrategy({
    required ShotContext context,
    required ShotPreferences profile,
    required Club recommendedClub,
  }) {
    var target = 'Center of green';
    var miss = 'Safe side of the green';
    var reasoning = '';

    // Aggressiveness
    switch (context.aggressiveness) {
      case Aggressiveness.conservative:
        target = 'Center of the green, away from trouble';
        reasoning = 'Conservative approach favors the widest part of the green.';
      case Aggressiveness.normal:
        target = 'Center of the green';
        reasoning = 'Standard approach targeting the middle of the green.';
      case Aggressiveness.aggressive:
        target = 'Pin location';
        reasoning = 'Aggressive play targeting the flag.';
    }

    // Miss tendency
    switch (profile.missTendency) {
      case MissTendency.right:
        miss = 'Favor the left side — your miss tends right';
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Accounting for right miss tendency.';
      case MissTendency.left:
        miss = 'Favor the right side — your miss tends left';
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Accounting for left miss tendency.';
      case MissTendency.thin:
        miss = 'Club up to account for thin contact tendency';
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Thin misses lose distance.';
      case MissTendency.fat:
        miss = 'Club up to account for heavy contact tendency';
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Heavy contact loses distance.';
      case MissTendency.straight:
        miss = 'No major miss adjustment needed';
    }

    // Slope-induced shape changes
    final clubShape = profile.stockShapeFor(recommendedClub);
    switch (context.slope) {
      case Slope.ballBelowFeet:
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Ball below feet promotes a fade.';
        if (clubShape == StockShape.fade) {
          miss = 'Favor the left side — stance will exaggerate your fade';
        }
      case Slope.ballAboveFeet:
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Ball above feet promotes a draw.';
        if (clubShape == StockShape.draw) {
          miss = 'Favor the right side — stance will exaggerate your draw';
        }
      // ignore: no_default_cases
      default:
        break;
    }

    // Crosswind
    switch (context.windDirection) {
      case WindDirection.crossLeftToRight:
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Crosswind L→R will push the ball right.';
        target += ', aiming slightly left to allow for wind';
      case WindDirection.crossRightToLeft:
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Crosswind R→L will push the ball left.';
        target += ', aiming slightly right to allow for wind';
      // ignore: no_default_cases
      default:
        break;
    }

    // Hazard notes
    final hazardNotes = context.hazardNotes.trim();
    if (hazardNotes.isNotEmpty) {
      final lower = hazardNotes.toLowerCase();
      if (lower.contains('water left') || lower.contains('hazard left')) {
        miss = 'Miss right — water left';
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Water left makes right the safe miss.';
      } else if (lower.contains('water right') ||
          lower.contains('hazard right')) {
        miss = 'Miss left — water right';
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Water right makes left the safe miss.';
      }
      if (lower.contains('bunker short')) {
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'Bunker short of green — make sure to take enough club.';
      }
      if (lower.contains('ob') || lower.contains('out of bounds')) {
        if (reasoning.isNotEmpty) reasoning += ' ';
        reasoning += 'OB in play — favor the safe side.';
      }
    }

    return TargetStrategy(
      target: target,
      preferredMiss: miss,
      reasoning: reasoning,
    );
  }

  // ── Describe adjustments (iOS lines 391-428) ────────────────────

  static List<String> describeAdjustments({
    required ShotContext context,
    IronType? ironType,
  }) {
    final out = <String>[];

    final wind = windAdjustment(
      strength: context.windStrength,
      direction: context.windDirection,
      baseDistance: context.distanceYards,
    );
    if (wind.abs() > 0.5) {
      final sign = wind > 0 ? '+' : '';
      out.add(
        'Wind (${_displayName(context.windStrength)} ${_displayName(context.windDirection)}): $sign${wind.round()} yards',
      );
    }

    final lie = lieAdjustment(context.lieType);
    if (lie.abs() > 0.5) {
      final sign = lie > 0 ? '+' : '';
      out.add(
        'Lie (${_displayName(context.lieType)}): $sign${lie.round()} yards',
      );
    }

    final slope = slopeAdjustment(
      slope: context.slope,
      baseDistance: context.distanceYards,
    );
    if (slope.abs() > 0.5) {
      final sign = slope > 0 ? '+' : '';
      out.add(
        'Slope (${_displayName(context.slope)}): $sign${slope.round()} yards',
      );
    }

    if (ironType != null) {
      final gi = ironTypeLieAdjustment(
        lieType: context.lieType,
        ironType: ironType,
      );
      if (gi.abs() > 0.5) {
        final shortName =
            ironType == IronType.superGameImprovement ? 'SGI' : 'GI';
        out.add(
          '$shortName iron penalty (${_displayName(context.lieType)}): +${gi.round()} yards',
        );
      }
    }

    if (out.isEmpty) {
      out.add('No adjustments — clean conditions');
    }
    return out;
  }

  // ── Display-name helpers ────────────────────────────────────────
  //
  // The iOS engine pulls `displayName` off each enum. We don't want
  // to bloat `golf_enums.dart` with display strings (those belong
  // in a UI layer), so we keep a small switch here for the
  // describeAdjustments output.

  static String _displayName(Object value) {
    if (value is WindStrength) {
      switch (value) {
        case WindStrength.none:
          return 'None';
        case WindStrength.light:
          return 'Light';
        case WindStrength.moderate:
          return 'Moderate';
        case WindStrength.strong:
          return 'Strong';
      }
    }
    if (value is WindDirection) {
      switch (value) {
        case WindDirection.into:
          return 'Into';
        case WindDirection.helping:
          return 'Helping';
        case WindDirection.crossLeftToRight:
          return 'Cross L→R';
        case WindDirection.crossRightToLeft:
          return 'Cross R→L';
      }
    }
    if (value is LieType) {
      switch (value) {
        case LieType.fairway:
          return 'Fairway';
        case LieType.firstCut:
          return 'First Cut';
        case LieType.rough:
          return 'Rough';
        case LieType.deepRough:
          return 'Deep Rough';
        case LieType.greensideBunker:
          return 'Greenside Bunker';
        case LieType.fairwayBunker:
          return 'Fairway Bunker';
        case LieType.hardpan:
          return 'Hardpan';
        case LieType.pineStraw:
          return 'Pine Straw';
        case LieType.treesObstructed:
          return 'Trees / Obstructed';
      }
    }
    if (value is Slope) {
      switch (value) {
        case Slope.level:
          return 'Level';
        case Slope.uphill:
          return 'Uphill';
        case Slope.downhill:
          return 'Downhill';
        case Slope.ballAboveFeet:
          return 'Ball Above Feet';
        case Slope.ballBelowFeet:
          return 'Ball Below Feet';
      }
    }
    return value.toString();
  }
}
